### PR TITLE
Standardize on Stack for backend methods.

### PR DESCRIPTION
**Pull request by pgavlin**
The current pattern of using backend.Stack values in the CLI but
accepting backend.StackReference values in backend methods leads to
frequent transitions between these types. In the case of the HTTP
backend, the transition from a StackReference to a Stack requires an API
call. These changes refactor the backend.Backend API such that most of
its methods accept backend.Stack values in place of
backend.StackReference values, which avoids these transitions.

This removes two calls to the getStack API on the startup path of a
`pulumi preview`.

### CODE REVIEW

**Comment by pgavlin** [comment]
> If we are constructing Stack objects unnecessarily, wouldn't we want to prune the places where we create those objects first? And instead just pass around a StackReference for as long as possible?

Either direction is a reasonable direction, and I considered them both. I chose this direction because it is more consistent with the design of the rest of the backend package as it exists today. As it is currently written--where the exported methods of everything besides the `Backend` interface work in terms of `Stack` values--the API encourages the use of `Stack` pretty much exclusively.

Moving to a `StackReference`-centric API is a much larger change, and comes at a bit more mental overhead (where do I actually need a `Stack` as opposed to a `StackReference`?). So I think this is the better approach for the time being.

File: pkg/backend/filestate/backend.go [hunk 1]
```diff
@@ -622,9 +598,9 @@ func GetLogsForTarget(target *deploy.Target, query operations.LogQuery) ([]opera
 }
 
 func (b *localBackend) ExportDeployment(ctx context.Context,
-	stackRef backend.StackReference) (*apitype.UntypedDeployment, error) {
+	stk backend.Stack) (*apitype.UntypedDeployment, error) {
```
Comment (justinvp) [inline pkg/backend/filestate/backend.go]: Nit: `stk` => `stack`?

File: pkg/backend/filestate/backend.go [hunk 2]
```diff
@@ -650,10 +626,10 @@ func (b *localBackend) ExportDeployment(ctx context.Context,
 	}, nil
 }
 
-func (b *localBackend) ImportDeployment(ctx context.Context, stackRef backend.StackReference,
+func (b *localBackend) ImportDeployment(ctx context.Context, stk backend.Stack,
```
Comment (justinvp) [inline pkg/backend/filestate/backend.go]: Nit: `stk` => `stack`?

File: pkg/backend/filestate/backend.go (hunk 1 above)
Comment (pgavlin) [reply pkg/backend/filestate/backend.go]: Can't: doing so shadows the `stack` package, and that package is required by this function.

File: pkg/backend/filestate/backend.go (hunk 2 above)
Comment (pgavlin) [reply pkg/backend/filestate/backend.go]: Same here.

**Pull request merged**